packages/react-aria-components/src/Collection.tsx

```diff
@@ -173,15 +173,18 @@ class BaseNode<T> {
     child.nextSibling = null;
     this.lastChild = child;
 
-    this.ownerDocument.dirtyNodes.add(this);
+    this.ownerDocument.markDirty(this);
     this.ownerDocument.addNode(child);
+    this.ownerDocument.endTransaction();
+    this.ownerDocument.queueUpdate();
   }
 
   insertBefore(newNode: ElementNode<T>, referenceNode: ElementNode<T>) {
     if (referenceNode == null) {
       return this.appendChild(newNode);
     }
 
+    this.ownerDocument.startTransaction();
```

Comment:
just to clarify, do we need this transaction? this whole block of code is not async, so nothing should be able to get between these two?